File: design_patterns_101/Additional Patterns/Chain of Responsibility Pattern/chainResponsibilityPattern.py
from __future__ import annotations
from abc import ABC, abstractmethod
from typing import Optional

# Handler interface
class Handler(ABC):
    @abstractmethod
    def set_next(self, handler: Handler) -> Handler:
        pass

    @abstractmethod
    def handle(self, request) -> Optional[str]:
        pass

# Base Handler
class AbstractHandler(Handler):
    _next_handler: Handler = None

    def set_next(self, handler: Handler) -> Handler:
        self._next_handler = handler
        return handler

    @abstractmethod
    def handle(self, request) -> Optional[str]:
        if self._next_handler:
            return self._next_handler.handle(request)
        return None

# Concrete Handlers
class ConcreteHandlerA(AbstractHandler):
    def handle(self, request) -> Optional[str]:
        if request == "A":
            return f"ConcreteHandlerA: Handled {request}"
        else:
            return super().handle(request)

class ConcreteHandlerB(AbstractHandler):
    def handle(self, request) -> Optional[str]:
        if request == "B":
            return f"ConcreteHandlerB: Handled {request}"
        else:
            return super().handle(request)

class ConcreteHandlerC(AbstractHandler):
    def handle(self, request) -> Optional[str]:
        if request == "C":
            return f"ConcreteHandlerC: Handled {request}"
        else:
            return super().handle(request)

# Usage
def client_code(handler: Handler) -> None:
    for request in ["A", "B", "C", "D"]:
        result = handler.handle(request)
        if result:
            print(result)
        else:
            print(f"No handler found for {request}")

handler_a = ConcreteHandlerA()
handler_b = ConcreteHandlerB()
handler_c = ConcreteHandlerC()

handler_a.set_next(handler_b).set_next(handler_c)

print("Chain: A -> B -> C")
client_code(handler_a)

File: design_patterns_101/Additional Patterns/Chain of Responsibility Pattern/chainResponsibilityPattern.js
// Handler interface
class Handler {
    setNext(handler) {
      throw new Error("Abstract method!");
    }
  
    handle(request) {
      throw new Error("Abstract method!");
    }
  }
  
  // Base Handler
  class AbstractHandler extends Handler {
    setNext(handler) {
      this.nextHandler = handler;
      return handler;
    }
  
    handle(request) {
      if (this.nextHandler) {
        return this.nextHandler.handle(request);
      }
      return null;
    }
  }
  
  // Concrete Handlers
  class ConcreteHandlerA extends AbstractHandler {
    handle(request) {
      if (request === "A") {
        return `ConcreteHandlerA: Handled ${request}`;
      }
      return super.handle(request);
    }
  }
  
  class ConcreteHandlerB extends AbstractHandler {
    handle(request) {
      if (request === "B") {
        return `ConcreteHandlerB: Handled ${request}`;
      }
      return super.handle(request);
    }
  }
  
  class ConcreteHandlerC extends AbstractHandler {
    handle(request) {
      if (request === "C") {
        return `ConcreteHandlerC: Handled ${request}`;
      }
      return super.handle(request);
    }
  }
  
  // Usage
  function clientCode(handler) {
    const requests = ["A", "B", "C", "D"];
    requests.forEach(request => {
      const result = handler.handle(request);
      if (result) {
        console.log(result);
      } else {
        console.log(`No handler found for ${request}`);
      }
    });
  }
  
  const handlerA = new ConcreteHandlerA();
  const handlerB = new ConcreteHandlerB();
  const handlerC = new ConcreteHandlerC();
  
  handlerA.setNext(handlerB).setNext(handlerC);
  
  console.log("Chain: A -> B -> C");
  clientCode(handlerA);
  